### PR TITLE
[IMP] mail: add support for guest users

### PR DESCRIPTION
**Pull request by oomsveta**
- Create new model `mail.guest` for guests.
 - Rewrite some RPCs to target routes rather than model methods so that
   guests are able to use them.
 - Patch JS and python models to support guests.
 - Create a stand-alone page and boot the channel in it.

task-2494829

### CODE REVIEW

File: addons/mail/models/mail_thread.py
Comment (mart-e) [inline addons/mail/models/mail_thread.py:1809]: passing a browserecord in the context is always a bit weird (and not rpc compatible), why not an id?

File: addons/mail/models/mail_thread.py
Comment (mart-e) [inline addons/mail/models/mail_thread.py:1844]: is this a recordset or an id?

File: addons/mail/controllers/discuss_controller.py
Comment (mart-e) [inline addons/mail/controllers/discuss_controller.py:255]: why not use browse().exists()?

File: addons/mail/models/mail_guest.py
Comment (mart-e) [inline addons/mail/models/mail_guest.py:38]: do you really need two cookies? isn't the token enough to retrieve an existing guest?

File: addons/mail/models/mail_guest.py
Comment (mart-e) [inline addons/mail/models/mail_guest.py:65]: `en_US` is not necessarily enabled on a db, you have `tools.misc.get_lang` that finds a valid lang for you

File: addons/mail/models/mail_guest.py
Comment (mart-e) [inline addons/mail/models/mail_guest.py:23]: why do you need a many2one? it seems you are only using it to retrieve the code so can just have a selection field `lang` as we do from time to time

File: addons/mail/models/mail_guest.py
Comment (mart-e) [inline addons/mail/models/mail_guest.py:87]: bad copy-paste

File: addons/mail/models/mail_guest.py
Comment (mart-e) [inline addons/mail/models/mail_guest.py:27]: why not use `name_get`?

File: addons/mail/models/res_users.py
Comment (mart-e) [inline addons/mail/models/res_users.py:98]: if there is nothing common, would be more readable in a traditional if/else multiline block

File: addons/bus/controllers/main.py
Comment (mart-e) [inline addons/bus/controllers/main.py:44]: if you are going to make it public, should add more verifications, not add a sudo

File: addons/mail/controllers/bus.py
Comment (mart-e) [inline addons/mail/controllers/bus.py:39]: there is a field `channel_ids` on `mail.guest`, can you use it?

File: addons/mail/models/mail_guest.py
Comment (mart-e) [inline addons/mail/models/mail_guest.py:71]: the only place you use that is to make a `guest.id` after, if you just return the id, you no longer need the `sudo(False)` above

File: addons/mail/models/mail_channel_invitation_link.py
Comment (mart-e) [inline addons/mail/models/mail_channel_invitation_link.py:9]: who can access this model? token field must be protected

**Review by mart-e — COMMENTED** [review]
first quick review

File: addons/mail/models/mail_thread.py
Comment (seb-odoo) [reply addons/mail/models/mail_thread.py:1809]: It's non-rpc compatible on purpose to prevent people from being able to set the value.

It's used in code where we want to actually know if current user is a guest, and we need to be able to trust it (not for security directly as access is checked before, but for example to tell when posting a new message if it should be considered posted as a specific guest, and which one).

File: addons/mail/controllers/discuss_controller.py
Comment (seb-odoo) [reply addons/mail/controllers/discuss_controller.py:255]: No specific reason, this could be changed as suggested.

File: addons/mail/models/mail_guest.py
Comment (seb-odoo) [reply addons/mail/models/mail_guest.py:38]: Isn't it less safe? I don't think psql will do a consteq equivalent when doing the search for it?

Also that would mean we need to index it on top of indexing the id, just to avoid one extra cookie that contains something as small as an id? I don't mind the change (if it's safe at least), I'm just wondering.

File: addons/mail/models/mail_guest.py
Comment (seb-odoo) [reply addons/mail/models/mail_guest.py:23]: copy/paste from website visitor is the reason

We'll look into changing it if you think there is a better way.

File: addons/mail/models/mail_guest.py
Comment (seb-odoo) [reply addons/mail/models/mail_guest.py:27]: It will likely contain more stuff in the result I think, need to double check after PR clean.

File: addons/mail/models/mail_guest.py
Comment (mart-e) [reply addons/mail/models/mail_guest.py:38]: Good point. We could use hash of tokens as we do in `res.users.apikeys._check_credentials` but it's more sensitive.

File: addons/mail/controllers/bus.py
Comment (sdegueldre) [inline addons/mail/controllers/bus.py:31]: Not sure why this was moved outside the if, partner_id is potentially not set here. Maybe the idea was to have this code in common between guests and users, in that case you should declare a variable (eg mail_channels) before the if-else, assign it in the if with this search, assign it in the else with your other search, and put the loop after the if-else.

File: addons/mail/controllers/discuss_controller.py
Comment (sdegueldre) [inline addons/mail/controllers/discuss_controller.py:50]: maybe it's just me but I find it weird that this function whose name starts with _get has an important side-effect such as setting cookies on the response, or that it takes a response as an argument in the first place when it's named "from_request".

Considering this is the only usage of that method, I would do the cookie setting in this controller instead.

File: addons/mail/controllers/discuss_controller.py
Comment (sdegueldre) [inline addons/mail/controllers/discuss_controller.py:184]: .search([('id', '=', thread_id)], limit=1) => .browse(thread_id)

**Review by sdegueldre — COMMENTED** [review]
Stopping my review here for today because I've got to go, already posting these comments so that you can take a look. Will continue tomorrow morning.

File: addons/mail/controllers/discuss_controller.py
Comment (xmo-odoo) [reply addons/mail/controllers/discuss_controller.py:184]: There is a small possible difference in semantics: `search` will return an empty recordset if the record is invalid. With `browse` you have to follow with `exists`.

**Review by xmo-odoo — CHANGES_REQUESTED** [review]

File: addons/mail/controllers/discuss_controller.py
Comment (xmo-odoo) [reply addons/mail/controllers/discuss_controller.py]: @seb-odoo a while ago irattachment grew a `raw` member which gives access to the raw payload of ir.attachment, rather than the base64-encoded `datas`.

So if you need or have base64 data it's fine to use `datas`, but if you have binary data (as here) all it does is add unnecessary overhead as the payload gets encoded, then immediately decoded, before being stored in the attachment store.

File: addons/mail/controllers/discuss_controller.py
Comment (rdeodoo) [reply addons/mail/controllers/discuss_controller.py]: It basically adds the current website in the context/request, which then change behavior of multiple methods (which depends of the context's webste, mostly attachment/views etc).

This is done through the `_add_dispatch_parameters()` [method](https://github.com/odoo/odoo/blame/master/addons/website/models/ir_http.py#L224) which is only done for [frontend requests](https://github.com/odoo/odoo/blame/master/addons/http_routing/models/ir_http.py#L383).

> I think it does smart stuff to try to find the correct country/lang/timezone/public partner/... based on website configuration if applicable/necessary. 

Indeed, and redirects accordingly, see [here](https://github.com/odoo/odoo/blame/master/addons/http_routing/models/ir_http.py#L493) (which is inside a `is_frontend` condition)

File: addons/web/models/ir_http.py
Comment (xmo-odoo) [inline addons/web/models/ir_http.py:181]: No that I'm thinking about it:

* get_resource_path shoudl just be `os.path.join`, everything it's going is redundant with `file_open`
* in fact I'm not sure working with list is useful at all as `/` as separator should work on all platform we support
* more importantly, `file_open` now has a `filter_ext` which is a list of acceptable extensions. It should very much be used here, even if the current callers look safe (as all current implementations of `_get_placeholder_filename` return static paths) I don't think there's any reason to allow other files than, say, png and jpg?

File: addons/mail/controllers/discuss_controller.py
Comment (seb-odoo) [reply addons/mail/controllers/discuss_controller.py]: I made a new task so we don't forget later: https://www.odoo.com/web#id=2635119&action=333&active_id=1519&model=project.task&view_type=form&cids=1&menu_id=4720

Marking as resolved for now.

File: addons/web/models/ir_http.py
Comment (seb-odoo) [reply addons/web/models/ir_http.py]: It was just a move in the first place, but I cleaned it as requested. Now just calling `file_open` with the string, and giving `filter_ext`.

File: addons/mail/models/mail_channel_partner.py
Comment (seb-odoo) [reply addons/mail/models/mail_channel_partner.py]: Finally done

File: addons/mail/controllers/discuss_controller.py
Comment (seb-odoo) [reply addons/mail/controllers/discuss_controller.py]: I guess I will remove it for now to avoid any surprising side effect.

File: addons/web/models/ir_http.py
Comment (xmo-odoo) [reply addons/web/models/ir_http.py]: :+1:

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

`_placeholder` is only invoked for the fallback of with the result of `_get_placeholder_filename` which are currently always literal strings.

Invocation has also been updated from the original version to filter on file extension for extra safety.

**Comment by seb-odoo** [comment]
I guess it's time to officially mark this one as ready for review. It will be merged today.

- Overall state should be pretty good.
- Still have one last todo in code that I plan to solve.
- Still have runbot to re-make green after last push.
- Still have to final review myself too.

File: addons/mail/views/discuss_public_templates.xml
Comment (alexkuhn) [inline addons/mail/views/discuss_public_templates.xml:50]: replace tabs by 4 whitespaces

**Review by alexkuhn — COMMENTED** [review]
LGTM

File: addons/mail/models/mail_guest.py
Comment (odony) [inline addons/mail/models/mail_guest.py:62]: use SKIP LOCKED instead of NOWAIT, and save yourself the trouble of savepoints and exception handling:

```suggestion
        query = """
            UPDATE mail_guest
            SET timezone = %s
            WHERE id IN (
                SELECT id FROM mail_guest WHERE id = %s
                FOR NO KEY UPDATE SKIP LOCKED
            )
        """
        self.env.cr.execute(query, (timezone, self.id))
```

Comment (odony) [reply addons/mail/models/mail_guest.py:62]: and yes, if you copy/pasted that, it should be fixed in website_visitor too

Comment (seb-odoo) [reply addons/mail/models/mail_guest.py:62]: I'll try it.

It's a direct copy from website visitor so we might want to change it there too then?

File: addons/mail/models/mail_channel.py
Comment (odony) [inline addons/mail/models/mail_channel.py:78]: why 8?? that's much less entropy than before (~ 122bit vs 45!) and it also fails to account for hard-to-distinguish characters (e.g. 1 vs I vs l vs 0 vs o vs O), if readability and length were factors.

You need 21 chars to keep the entropy (or 22 if you use a hardcoded list of easy-to-read chars).

Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:78]: The goal is to have a rather short URL to share. It's 8 char for a functional reason because that's how many are on share link in Discord.

Good point to exclude hard-to-distinguish but then it's even worse in terms of security...

Ideally I'd like that we generate the uuid only when we actually need to share it, but some other parts of the code still rely on it being always set for now...

Do you have any suggestion with this in mind?
Do we actually need as much entropy as before, or can we settle for a middle ground?

Maybe at least log something in case of wrong value in the invite route so that fail2ban can be set to prevent the most simple brute force?

Note: if this is not negotiable I'll put 22, I don't want to block the PR for this.

Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:78]: Renaming is too much work (and risk of breaking other stuff) for the time I have available now, but ok to improve after v15. This PR opens a lot of clean up opportunities also in portal chatter and livechat.

File: addons/mail/controllers/discuss.py
Comment (odony) [inline addons/mail/controllers/discuss.py:40]: those are meant to be session cookies? can this be explicitly documented?
also, could we make them `httponly=True` since they never need to be read from the JS side

ideally they should be `secure=True` too, but we cannot 100% force HTTPS-only deployments, I think

File: addons/mail/controllers/discuss.py
Comment (odony) [inline addons/mail/controllers/discuss.py:20]: shouldn't this be better namespaced?

```suggestion
    @http.route('/chat/join/<int:channel_id>/<string:invitation_token>', methods=['GET'], type='http', auth='public')
```

File: addons/mail/controllers/discuss.py
Comment (odony) [inline addons/mail/controllers/discuss.py:24]: is there no limitation on which kind of channel you can join with this? It seems this should not be possible for livechats, nor for other private channels?

File: addons/mail/controllers/discuss.py
Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:20]: The goal is to have the shortest share URL possible. Note that I'm expecting this URL might change again soon depending on feedback that will inevitably come when the oxp videos will be made and people start to test it more seriously.

File: addons/mail/controllers/discuss.py
Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:24]: In terms of security if you have the uuid you can already sneak in and receive messages for all kind of channels before this PR (support for uuid bus notif + if livechat is actually installed there are routes to fetch all messages based on uuid).
This new route just makes people using it formally a member of said channel.

But it would make sense to at least prevent it for channel of type chat, as those are exclusively between 2 persons are not supposed to be joined in any way.

Private channels (of channel type group in particular) are the main target of this PR, the spec is to create a private group on the fly and invite people with a link.

"Public" channel (which is currently equals to visible to all internal users) is actually the least likely to be shared this way.

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:24]: As for livechat it's actually common enough for a livechat operator to request help from his colleague on some discussion, in which case being able to share the link with him is more convenient than the (already long time existing) invite feature by name.

File: addons/mail/controllers/discuss.py
Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:40]: Indeed, those can be considered like session cookie for guest. Maybe some day we can merge it with the real session and maybe also with website visitor? But it's out of scope for now.

Good idea to add `httponly=True`, we don't need those cookies in JS.

File: addons/mail/models/mail_channel.py
Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:48]: Isn't `choices` from `random` instead of from `secrets`, so not safe for security purposes?

File: addons/mail/controllers/discuss.py
Comment (odony) [reply addons/mail/controllers/discuss.py]: > Indeed, those can be considered like session cookie for guest. Maybe some day we can merge it with the real session and maybe also with website visitor? But it's out of scope for now.

Wait, I meant "session cookie" as in "non-permanent cookies that will be discarded if you close your browser". Because they have no expiration date. This is different from typical expiration for our "session_id" cookie ;-)

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py]: Then I guess we don't want them to be "session", if the guest is closing his browser it would be convenient to restore his identity instead of creating a new one. So we would want those cookie to live as long as possible.

Is there anything to change here then? In code or in the comment I added?

File: addons/mail/models/mail_channel.py
Comment (seb-odoo) [reply addons/mail/models/mail_channel.py]: Made it to be 10 with the provided list.

File: addons/mail/models/mail_channel_partner.py
Comment (seb-odoo) [reply addons/mail/models/mail_channel_partner.py:33]: Good to know, I didn't know about those tools.

If you think it is worth it feel free to improve the tool and edit them all after this PR is merged (because there are few other occurrences of such index in the rest of the code too).

**Review by seb-odoo — COMMENTED** [review]
note to self

**Review by seb-odoo — APPROVED** [review]
Hoping it will still be green and no other PR introduces a conflict until then.

@robodoo r+

File: addons/mail/controllers/discuss.py
Comment (odony) [reply addons/mail/controllers/discuss.py]: > Is there anything to change here then? In code or in the comment I added?

Well yes, you have to set an expiration date or max-age on the cookie, that's how you make it permanent on the browser side. For `session_id` we put the validity [at 3 months](https://github.com/odoo/odoo/blob/master/odoo/http.py#L1429), but we auto-renew the cookie frequently.
For `visitor_uuid` we use a fixed [1 year validity](https://github.com/odoo/odoo/blob/master/addons/website/models/website_visitor.py#L205-L206).

So it depends: how long should we keep the same "guest"? 3 months or 1 year seems plausible values.

Then just pass an appropriate max_age value to set_cookie, as in the examples.

File: addons/mail/controllers/discuss.py
Comment (odony) [inline addons/mail/controllers/discuss.py:44]: One more thing - proliferation of cookies is bad taste, and with GDPR every single one of them needs to be documented for compliance reasons, with the purpose, etc.

So it would be better to use a structured cookie instead of 2 different ones. It's also nice to save some bytes by not making the cookie name longer than strictly necessary. Remember that this stuff is sent for *every single request* that will be made by the browser. It's not like we need users to be able to read the cookies.

```suggestion
            # dgid = "discord guest id"
            response.set_cookie('dgid', f"{guest.id}|{guest.access_token}", httponly=True)
```

then split it when you need the values...

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:44]: Ok for the suggestions, I'll make the changes to cookie post merge, or directly if we need to fix conflict at some point with all that's pending on mergebot.

Comment (seb-odoo) [reply addons/mail/controllers/discuss.py:44]: https://github.com/odoo/odoo/pull/75888

File: addons/mail/models/mail_channel.py
Comment (xmo-odoo) [reply addons/mail/models/mail_channel.py]: Gah, I hate when I miss imports. Yes, at least when using the toplevel function (`secrets.choice` is just an alias to a `SystemRandom().choice`, so you could do that for a secure version of `choices`).

**Pull request closed without merging**
